tests/test_periodic_tasks.cpp

```diff
@@ -0,0 +1,108 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <gtest/gtest.h>
+
+#include <bm/bm_sim/periodic_task.h>
+
+#include <chrono>
+#include <functional>
+
+using namespace bm;
+
+class PeriodicIncrementor {
+ public:
+  PeriodicIncrementor(const std::chrono::milliseconds &interval, int *i)
+      : i(i), increment_task("increment",
+                             std::bind(&PeriodicIncrementor::increment, this),
+                             interval) {}
+
+  void increment() {
+    (*i)++;
+  }
+  int *i;
+
+ private:
+  PeriodicTask increment_task;
+};
+
+static constexpr int kSleepTimeMs = 500;
+static constexpr int kPeriods = 5;
+
+static constexpr std::chrono::milliseconds kInterval{kSleepTimeMs / kPeriods};
+static constexpr std::chrono::milliseconds kSleepTime{kSleepTimeMs};
+
+// Tasks registered before the start of the thread execute on start
+TEST(PeriodicExtern, PreStartRegistration) {
+  int i = 0;
+  PeriodicIncrementor incrementor(kInterval, &i);
+  std::this_thread::sleep_for(kSleepTime);
+
+  ASSERT_EQ(i, 0);
```

Comment:
I didn't know that when I started writing GTests, but unless a violation of the expectation causes the rest to crash, I believe `EXPECT_*` macros are preferred.